File: src/rag_harness.py
"""
X_Ray RUST CONVERSION TEST HARNESS
════════════════════════════════════════════════════════════════════

Comprehensive automated testing for:
1. Crash prevention and error handling
2. Real website RAG pipeline
3. Performance profiling (Python vs Rust)
4. X_Ray analysis on converted code
5. Question answering from RAG

Websites tested:
- Crestafund.com (crowdfunding platform)
- Oradea.ro (municipal website)
"""

import sys
import os
import time
import json
import traceback
import psutil
from typing import Dict, Tuple, Optional
from datetime import datetime

# Add RAG_RAT to path
sys.path.insert(0, os.path.dirname(__file__))

print("\n" + "=" * 80)
print("RUST CONVERSION REAL-WORLD TEST HARNESS")
print("=" * 80)

# ═══════════════════════════════════════════════════════════════════════════
# CRASH PREVENTION LAYER
# ═══════════════════════════════════════════════════════════════════════════


class SafeImporter:
    """Safely import modules and catch all errors."""

    def __init__(self):
        self.errors = []
        self.warnings = []
        self.imports = {}

    def safe_import(self, module_name: str, display_name: str = None) -> bool:
        """Try to import a module, catch and log errors."""
        display_name = display_name or module_name
        try:
            self.imports[module_name] = __import__(module_name)
            print(f"  ✓ {display_name:40} success")
            return True
        except Exception as e:
            error_msg = f"{display_name}: {str(e)[:100]}"
            self.errors.append(error_msg)
            print(f"  ✗ {display_name:40} FAILED: {str(e)[:50]}")
            return False

    def report(self):
        """Print import summary."""
        print("\n" + "─" * 80)
        print("IMPORT VERIFICATION")
        print("─" * 80)
        print(f"Successful: {len(self.imports)} modules")
        print(f"Failed: {len(self.errors)} modules")
        if self.errors:
            print("\nErrors:")
            for err in self.errors:
                print(f"  • {err}")
                pass
        return len(self.errors) == 0


# ═══════════════════════════════════════════════════════════════════════════
# INITIALIZATION
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "─" * 80)
print("PHASE 1: SAFE INITIALIZATION")
print("─" * 80)

importer = SafeImporter()

# Critical imports
importer.safe_import("requests", "requests (HTTP)")
importer.safe_import("bs4", "BeautifulSoup (HTML parsing)")
importer.safe_import("psutil", "psutil (System monitoring)")

# RAG_RAT imports
importer.safe_import("content_extractor", "Content Extractor")
importer.safe_import("content_extractor_rust_bridge", "Content Extractor Rust Bridge")
importer.safe_import("inference_guard_rust_bridge", "Inference Guard Rust Bridge")

# Check Rust module
try:
    import rag_rat_rust

    RUST_AVAILABLE = True
    importer.imports["rag_rat_rust"] = rag_rat_rust
    print(f"  ✓ {'Rust Extension Module (rag_rat_rust.pyd)':40} success")
except ImportError as e:
    RUST_AVAILABLE = False
    importer.warnings.append(f"Rust module not available: {e}")
    print(f"  ⚠ {'Rust Extension Module (rag_rat_rust.pyd)':40} not available")
if not importer.report():
    print("\n❌ Critical imports failed. Exiting.")
    sys.exit(1)

print("\n✅ All critical modules loaded successfully")
if RUST_AVAILABLE:
    print("✅ Rust acceleration available")
else:
    print("⚠️  Running Python-only mode")

# ═══════════════════════════════════════════════════════════════════════════
# WEBSITE FETCHING WITH CRASH PREVENTION
# ═══════════════════════════════════════════════════════════════════════════


def safe_fetch_website(url: str, timeout: int = 30) -> Tuple[bool, str, Optional[str], Optional[int]]:
    """
    Safely fetch website content with error handling.

    Returns: (success, content, error_message, status_code)
    """
    try:
        print(f"\n  📥 Fetching {url}...")
        import requests

        response = requests.get(url, timeout=timeout)
        response.raise_for_status()

        print(f"     Status: {response.status_code}")
        print(f"     Size: {len(response.content) / 1024:.1f} KB")
        return True, response.content, None, response.status_code

    except requests.exceptions.Timeout:
        error = f"Timeout after {timeout}s"
        print(f"     ❌ {error}")
        return False, "", error, None

    except requests.exceptions.ConnectionError as e:
        error = f"Connection error: {str(e)[:50]}"
        print(f"     ❌ {error}")
        return False, "", error, None

    except requests.exceptions.HTTPError as e:
        status = e.response.status_code if e.response else None
        error = f"HTTP {status}: {str(e)[:50]}"
        print(f"     ❌ {error}")
        return False, "", error, status

    except Exception as e:
        error = f"Unexpected error: {str(e)[:50]}"
        print(f"     ❌ {error}")
        print(f"     {traceback.format_exc()[:200]}")
        return False, "", error, None


# ═══════════════════════════════════════════════════════════════════════════
# SAFE CONTENT PROCESSING
# ═══════════════════════════════════════════════════════════════════════════


def safe_process_content(html_content: bytes, url: str) -> Dict:
    """Safely process HTML content with crash prevention."""

    result = {
        "url": url,
        "success": False,
        "html_size": len(html_content),
        "text_extracted": 0,
        "images_found": 0,
        "text_preview": "",
        "errors": [],
        "warnings": [],
    }

    try:
        # Decode HTML
        try:
            from content_extractor_rust_bridge import decode_response_text

            text = decode_response_text(html_content, None)
            result["decode_method"] = "Rust"
        except Exception as e:
            result["warnings"].append(f"Rust decode failed, using Python: {str(e)[:50]}")
            try:
                text = html_content.decode("utf-8", errors="replace")
            except Exception:
                text = html_content.decode("iso-8859-1", errors="replace")
            result["decode_method"] = "Python fallback"

        result["text_extracted"] = len(text)
        result["text_preview"] = text[:200]

        # Extract images
        try:
            from content_extractor_rust_bridge import extract_images

            images = extract_images(text, url)
            result["images_found"] = len(images)
        except Exception as e:
            result["warnings"].append(f"Image extraction failed: {str(e)[:50]}")
            result["images_found"] = 0

        result["success"] = True

    except Exception as e:
        result["errors"].append(f"Content processing failed: {str(e)[:100]}")
        result["success"] = False
        print(f"\n     ❌ Content processing error: {str(e)}")
        traceback.print_exc()

    return result


# ═══════════════════════════════════════════════════════════════════════════
# PHASE 2: WEBSITE SCANNING
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "=" * 80)
print("PHASE 2: SAFE WEBSITE SCANNING")
print("=" * 80)

websites = {
    "crestafund": "https://www.crestafund.com",
    "oradea": "https://oradea.ro",
}

website_data = {}

for name, url in websites.items():
    print(f"\n{name.upper()}")
    print("─" * 80)

    # Fetch
    success, content, error, status = safe_fetch_website(url)

    if not success:
        print(f"  ❌ Failed to fetch {url}")
        website_data[name] = {
            "url": url,
            "success": False,
            "error": error,
            "status_code": status,
        }
        continue

    # Process
    result = safe_process_content(content, url)
    website_data[name] = result

    print(f"  ✓ HTML decoded: {result['text_extracted']:,} characters")
    print(f"  ✓ Images found: {result['images_found']}")
    print(f"  ✓ Decode method: {result.get('decode_method', 'unknown')}")
    if result.get("errors"):
        print(f"  ⚠ Errors: {result['errors']}")
        pass
    if result.get("warnings"):
        print(f"  ⚠ Warnings: {result['warnings']}")
        pass
        # ═══════════════════════════════════════════════════════════════════════════
        # PHASE 3: CRASH DETECTION & REPORTING
        # ═══════════════════════════════════════════════════════════════════════════

        pass
print("\n" + "=" * 80)
print("PHASE 3: CRASH DETECTION REPORT")
print("=" * 80)

crashes = []
for name, data in website_data.items():
    if not data.get("success", True):
        crashes.append({"site": name, "reason": data.get("error", "Unknown error")})

if crashes:
    print(f"\n❌ {len(crashes)} CRASHES DETECTED:")
    for crash in crashes:
        print(f"  • {crash['site']}: {crash['reason']}")
        pass
else:
    print("\n✅ NO CRASHES - All websites processed safely!")

# ═══════════════════════════════════════════════════════════════════════════
# PHASE 4: BUILD RAG KNOWLEDGE BASE
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "=" * 80)
print("PHASE 4: BUILD RAG KNOWLEDGE BASE")
print("=" * 80)

# For now, create a simple knowledge base from extracted text
rag_knowledge = {}

for name, data in website_data.items():
    if data.get("success") and data.get("text_extracted", 0) > 0:
        rag_knowledge[name] = {
            "url": data["url"],
            "content": data.get("text_preview", "")[:500],  # Store preview
            "size": data["text_extracted"],
            "images": data["images_found"],
        }
        print(f"✓ {name}: Added {data['text_extracted']:,} chars to RAG")
print(f"\n✅ RAG Knowledge Base Ready: {len(rag_knowledge)} sources")
# ═══════════════════════════════════════════════════════════════════════════
# PHASE 5: SAMPLE QUESTIONS
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "=" * 80)
print("PHASE 5: QUESTION ANSWERING FROM RAG")
print("=" * 80)

sample_questions = [
    "What is Oradea and what are its main features?",
    "What kind of organization is Crestafund?",
    "What services or content are mentioned on these websites?",
]

print("\nSample questions to answer from RAG:")
for i, q in enumerate(sample_questions, 1):
    print(f"  {i}. {q}")
    pass
print("\n(Full LLM integration would complete the RAG pipeline)")

# ═══════════════════════════════════════════════════════════════════════════
# PHASE 6: PERFORMANCE PROFILING
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "=" * 80)
print("PHASE 6: PERFORMANCE PROFILING")
print("=" * 80)

process = psutil.Process()

total_memory_start = process.memory_info().rss / 1024 / 1024
total_time_start = time.time()

print(f"\nMemory usage: {total_memory_start:.2f} MB")
print(f"CPU percent: {process.cpu_percent(interval=1):.1f}%")
# ═══════════════════════════════════════════════════════════════════════════
# PHASE 7: SUMMARY REPORT
# ═══════════════════════════════════════════════════════════════════════════

print("\n" + "=" * 80)
print("PHASE 7: TEST SUMMARY")
print("=" * 80)

total_time = time.time() - total_time_start
total_memory = process.memory_info().rss / 1024 / 1024

summary = {
    "timestamp": datetime.now().isoformat(),
    "rust_available": RUST_AVAILABLE,
    "tests_run": len(website_data),
    "tests_passed": sum(1 for d in website_data.values() if d.get("success", False)),
    "crashes": len(crashes),
    "total_content_size": sum(d.get("text_extracted", 0) for d in website_data.values()),
    "total_images": sum(d.get("images_found", 0) for d in website_data.values()),
    "total_time_seconds": total_time,
    "memory_mb": total_memory,
}

print("\n📊 TEST RESULTS:")
print(f"  ✓ Tests passed: {summary['tests_passed']}/{summary['tests_run']}")
print(f"  ✓ Crashes: {summary['crashes']}")
print(f"  ✓ Total content: {summary['total_content_size']:,} chars")
print(f"  ✓ Total images: {summary['total_images']}")
print(f"  ✓ Time: {summary['total_time_seconds']:.2f}s")
print(f"  ✓ Memory: {summary['memory_mb']:.2f} MB")
print(f"  ✓ Rust available: {'YES' if summary['rust_available'] else 'NO'}")
if summary["crashes"] == 0:
    print("\n✅ ALL TESTS PASSED - NO CRASHES!")
else:
    print(f"\n⚠️  {summary['crashes']} CRASH(ES) DETECTED")
    pass
    # ═══════════════════════════════════════════════════════════════════════════
    # SAVE RESULTS
    # ═══════════════════════════════════════════════════════════════════════════

    pass
results_file = "rag_harness_results.json"
with open(results_file, "w") as f:
    json.dump(
        {
            "summary": summary,
            "website_data": website_data,
            "crashes": crashes,
        },
        f,
        indent=2,
    )

print(f"\n💾 Results saved to: {results_file}")
print("\n" + "=" * 80)
print("✅ HARNESS COMPLETE")
print("=" * 80)

File: rust_output/src/rag_harness.rs
/// X_Ray RUST CONVERSION TEST HARNESS
/// ════════════════════════════════════════════════════════════════════
/// 
/// Comprehensive automated testing for:
/// 1. Crash prevention and error handling
/// 2. Real website RAG pipeline
/// 3. Performance profiling (Python vs Rust)
/// 4. X_Ray analysis on converted code
/// 5. Question answering from RAG
/// 
/// Websites tested:
/// - Crestafund.com (crowdfunding platform)
/// - Oradea.ro (municipal website)

use anyhow::{Result, Context};
use std::collections::HashMap;

pub static IMPORTER: std::sync::LazyLock<SafeImporter> = std::sync::LazyLock::new(|| Default::default());

pub static WEBSITES: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub static WEBSITE_DATA: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub static CRASHES: std::sync::LazyLock<Vec<serde_json::Value>> = std::sync::LazyLock::new(|| Vec::new());

pub static RAG_KNOWLEDGE: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub static SAMPLE_QUESTIONS: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

pub static PROCESS: std::sync::LazyLock<String /* psutil.Process */> = std::sync::LazyLock::new(|| Default::default());

pub const TOTAL_MEMORY_START: &str = "process.memory_info().rss / 1024 / 1024";

pub static TOTAL_TIME_START: std::sync::LazyLock<String /* time::time */> = std::sync::LazyLock::new(|| Default::default());

pub const TOTAL_TIME: &str = "std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - total_time_start";

pub const TOTAL_MEMORY: &str = "process.memory_info().rss / 1024 / 1024";

pub static SUMMARY: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub const RESULTS_FILE: &str = "rag_harness_results.json";

/// Safely import modules and catch all errors.
#[derive(Debug, Clone)]
pub struct SafeImporter {
    pub errors: Vec<serde_json::Value>,
    pub warnings: Vec<serde_json::Value>,
    pub imports: HashMap<String, serde_json::Value>,
}

impl SafeImporter {
    pub fn new() -> Self {
        Self {
            errors: vec![],
            warnings: vec![],
            imports: HashMap::new(),
        }
    }
    /// Try to import a module, catch and log errors.
    pub fn safe_import(&mut self, module_name: String, display_name: String) -> Result<bool> {
        // Try to import a module, catch and log errors.
        let mut display_name = (display_name || module_name);
        // try:
        {
            self.imports[module_name] = __import__(module_name);
            println!("  ✓ {:40} success", display_name);
            true
        }
        // except Exception as e:
    }
    /// Print import summary.
    pub fn report(&mut self) -> () {
        // Print import summary.
        println!("{}", ("\n".to_string() + ("─".to_string() * 80)));
        println!("{}", "IMPORT VERIFICATION".to_string());
        println!("{}", ("─".to_string() * 80));
        println!("Successful: {} modules", self.imports.len());
        println!("Failed: {} modules", self.errors.len());
        if self.errors {
            println!("{}", "\nErrors:".to_string());
            for err in self.errors.iter() {
                println!("  • {}", err);
                // pass
            }
        }
        self.errors.len() == 0
    }
}

/// Safely fetch website content with error handling.
/// 
/// Returns: (success, content, error_message, status_code)
pub fn safe_fetch_website(url: String, timeout: i64) -> Result<(bool, String, Option<String>, Option<i64>)> {
    // Safely fetch website content with error handling.
    // 
    // Returns: (success, content, error_message, status_code)
    // try:
    {
        println!("\n  📥 Fetching {}...", url);
        // TODO: import requests
        let mut response = /* reqwest::get( */&url).cloned().unwrap_or(/* timeout= */ timeout);
        response.raise_for_status();
        println!("     Status: {}", response.status_code);
        println!("     Size: {:.1} KB", (response.content.len() / 1024));
        (true, response.content, None, response.status_code)
    }
    // except requests.exceptions::Timeout as _e:
    // except requests.exceptions::ConnectionError as e:
    // except requests.exceptions::HTTPError as e:
    // except Exception as e:
}

/// Safely process HTML content with crash prevention.
pub fn safe_process_content(html_content: Vec<u8>, url: String) -> Result<HashMap> {
    // Safely process HTML content with crash prevention.
    let mut result = HashMap::from([("url".to_string(), url), ("success".to_string(), false), ("html_size".to_string(), html_content.len()), ("text_extracted".to_string(), 0), ("images_found".to_string(), 0), ("text_preview".to_string(), "".to_string()), ("errors".to_string(), vec![]), ("warnings".to_string(), vec![])]);
    // try:
    {
        // try:
        {
            // TODO: from content_extractor_rust_bridge import decode_response_text
            let mut text = decode_response_text(html_content, None);
            result["decode_method".to_string()] = "Rust".to_string();
        }
        // except Exception as e:
        result["text_extracted".to_string()] = text.len();
        result["text_preview".to_string()] = text[..200];
        // try:
        {
            // TODO: from content_extractor_rust_bridge import extract_images
            let mut images = extract_images(text, url);
            result["images_found".to_string()] = images.len();
        }
        // except Exception as e:
        result["success".to_string()] = true;
    }
    // except Exception as e:
    Ok(result)
}
